### ISSUE TITLE
[Bug]: Bluetooth permission on iOS permanently denied

### ISSUE THREAD

**Issue by lazarvgd** [Baseflow/flutter-permission-handler#1333]
### Please check the following before submitting a new issue.

- [X] I have searched the [existing issues](https://github.com/baseflow/flutter-permission-handler/issues).
- [X] I have carefully [read the documentation](https://github.com/Baseflow/flutter-permission-handler/blob/main/permission_handler/README.md) and verified I have added the required platform specific configuration.

### Please select affected platform(s)

- [ ] Android
- [X] iOS
- [ ] Windows

### Steps to reproduce

1. add the dependency to pubspec.yaml
2. add permission for BT in info.plist
3. update podfile with the following code:

```
      config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= [
              '$(inherited)',

              ## dart: PermissionGroup.bluetooth
              'PERMISSION_BLUETOOTH=1',

            ]
```

4. request permission in the code as following:

```
  @override
  Future<bool> getBluetoothPermission() async {
   final PermissionStatus  status = await Permission.bluetoothConnect.request();
    return status == PermissionStatus.granted;
  }
```

### Expected results

When the bt is enabled in Settings-> <app_name> -> Bluetooth, result of the code should be "granted", when it is disabled it should be some variant of denied.

### Actual results

The actual result is that the code always return **permanentlyDenied**.

### Code sample

<details><summary>Code sample</summary>

```dart
  @override
  Future<bool> getBluetoothPermission() async {
   final PermissionStatus  status = await Permission.bluetoothConnect.request();
    return status == PermissionStatus.granted;
  }
```

</details>

### Screenshots or video

<details>
<summary>Screenshots or video demonstration</summary>

![image](https://github.com/Baseflow/flutter-permission-handler/assets/11629801/fbc3c3d9-c3cf-4284-bb78-13a8557fa366)
![Screenshot 2024-06-10 at 10 21 53](https://github.com/Baseflow/flutter-permission-handler/assets/11629801/bfc83611-81c8-4c19-aea8-59e156bcfa90)

</details>

### Version

11.3.1

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.22.1, on macOS 14.5 23F79 darwin-arm64, locale en-GB)
    • Flutter version 3.22.1 on channel stable at /Users/lj/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a14f74ff3a (3 weeks ago), 2024-05-22 11:08:21 -0500
    • Engine revision 55eae6864b
    • Dart version 3.4.1
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/lj/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.89.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.88.0

[✓] Connected device (5 available)
    • SM S911B (mobile)               • RFCW202MYCV               • android-arm64  • Android 14 (API 34)
    • iPhone-J1HPW2QJMW (mobile)      • 00008110-001E14E90E3A401E • ios            • iOS 17.5.1 21F90
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS 14.5 23F79 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin         • macOS 14.5 23F79 darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google Chrome 125.0.6422.142

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

**Comment by ritheshSalyan** [comment]
I too had this issue.

The problem in my case was, that I had not run `pod install` after modifying `podfile`.

Solution:  reinstall the pod.

1. Move to ios folder
`cd ios`
2. Run the command `pod install`.

this solved my issue.

**Comment by lazarvgd** [comment]
The proposed solution does not work.

**Comment by TimHoogstrate** [comment]
Dear @lazarvgd,

Does this also happen while you test this in the Example app? 

Kins regards,

**Comment by tilongzs** [comment]
I have encountered the same problem.Sorry, I am not good at iOS compilation and the Example app cannot be compiled successfully.
Tested on iOS 16.7.7, iphone 8.

**Comment by Phenek** [comment]
Specific App Settings Permission
<img src="https://github.com/Baseflow/flutter-permission-handler/assets/4518686/5b0ce572-9f0e-429b-8ba0-abd35d642ddc" alt="image" width="300"/>

All this permissions returns `permanentlyDenied` status
````
      var status1 = await Permission.bluetooth.request();
      var status2 = await Permission.bluetoothAdvertise.request();
      var status3 = await Permission.bluetoothConnect.request();
      var status4 = await Permission.bluetoothScan.request();
````

In fact Permission is granted, I can scan nearby bluetooth devices!

**Comment by TimHoogstrate** [comment]
> I have encountered the same problem.Sorry, I am not good at iOS compilation and the Example app cannot be compiled successfully. Tested on iOS 16.7.7, iphone 8.

The example project compiles correctly.

@Phenek,

Please file a separate issue. Also check the correct plist entries and the pod file entries. Then run `flutter build ios`.

Kind regards,

**Comment by fufylev** [comment]
same issue for me -  spent already many hrs on this but still fail to solve the issue. A always get `Permission.denied` when in fact on the devise the permission is granted.

Any updates on this?
Flutter 3.22.2
iOS 17.6

**Comment by TimHoogstrate** [comment]
Dear @fufylev,

Please file a separate issue. And follow the instructions in the README. You probably forgot to update the podfile or the plist entries. 

Kind regards,

**Comment by Hoai-Phong** [comment]
@fufylev  
When I have granted permission on the app, why does it still show as permanentlyDenied?
Is there any way to fix it?
![image](https://github.com/user-attachments/assets/9ccc14dd-8a58-458d-8333-7aa4ccfd7e1f)

![image](https://github.com/user-attachments/assets/8565a36b-3d8a-4f19-a7dc-b610caba2612)

 permission_handler: ^11.3.1

**Comment by Phenek** [comment]
@TimHoogstrate  Sorry, but why do I need to file a separate issue?
This issue is what it is. No need to duplicate, don't you think?

My issue is exactly the title of this one.
Bluetooth permission on iOS permanently denied, even when in fact on the devise the permission is granted.

So I tried to put the project as code reference in pubspec.yaml

```
  permission_handler:
    path: ../../flutter-permission-handler/permission_handler
```

but I could not debug, because breakpoints don't stop on my Visual Studio Code. I don't why?
Can you please, share you debug advice @TimHoogstrate

**Comment by sahharYoucef** [comment]
Same issue on ios18 , flutter 3.22.2 with notifications permission

**Comment by djamel-madani-dev** [comment]
I'm facing the same issue !

**Comment by maxindieflow** [comment]
Facing same issue with notifications permission on iOS devices, flutter 3.22.4. Podfile and Info.plist is set up according to readme.

**Comment by Daeon97** [comment]
Hey everyone. I am also encountering this issue. Here let me share screenshots:

I have modified my `Podfile` with the appropriate macros accordingly

<img width="584" alt="Screenshot 2024-12-12 at 6 49 05 PM" src="https://github.com/user-attachments/assets/56cd4fc5-747e-4c49-b155-84bc87422b24" />

I have added appropriate descriptions in `Info.plist`
```
<key>NSBluetoothPeripheralUsageDescription</key>
<string>...</string>
<key>NSBluetoothAlwaysUsageDescription</key>
<string>...</string>
```

Here are the relevant contents of my `pubspec.yaml` and `pubspec.lock` files

`pubspec.yaml`
<img width="309" alt="Screenshot 2024-12-12 at 7 04 16 PM" src="https://github.com/user-attachments/assets/83d47ce6-6809-4000-b931-ff59d0c3fe66" />

`pubspec.lock`
<img width="692" alt="Screenshot 2024-12-12 at 7 03 56 PM" src="https://github.com/user-attachments/assets/5d8b7926-0d45-4488-8b92-4992de58c47e" />

Now I run my app. It actually asks for bluetooth permission and i grant it

https://github.com/user-attachments/assets/ebd5c05a-692b-4b5b-b753-6bd07c98daf0

https://github.com/user-attachments/assets/a760b409-675b-4a8f-bd3f-f1d7def836fb

You can clearly see that bluetooth permission is granted here
![IMG_1842](https://github.com/user-attachments/assets/6596d005-6996-473a-93a6-dd37a4c6b24e)

So why does this still happen?
<img width="944" alt="Screenshot 2024-12-12 at 7 17 03 PM" src="https://github.com/user-attachments/assets/da6a4a3f-d73b-4a55-a13d-e6e809536881" />

**Comment by Daeon97** [comment]
Okay so I have fixed this by eliminating the line `if target.name == "geolocator_apple"` and including a new line and the comment `## dart: PermissionGroup.bluetooth` just before the line `'PERMISSION_BLUETOOTH=1',`

<img width="582" alt="Screenshot 2024-12-12 at 8 42 00 PM" src="https://github.com/user-attachments/assets/bd86a35c-3b3c-4375-87cc-ebff045f539e" />

Specifically instead of

```
if target.name == "geolocator_apple"
  target.build_configurations.each do |config|
    config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= [
      '$(inherited)',
      'PERMISSION_BLUETOOTH=1',
    ]
  end
end
```

I now have

```
target.build_configurations.each do |config|
  config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= [
    '$(inherited)',

    ## dart: PermissionGroup.bluetooth
    'PERMISSION_BLUETOOTH=1',
  ]
end
```

**Comment by lazarvgd** [comment]
> Dear @lazarvgd,
> 
> Does this also happen while you test this in the Example app?
> 
> Kins regards,

I want to apologize for the late reply. 

The issue was happening in the example app. I couldn't wait for the fix so I ended writing my own permission handler.
I can confirm that the[ tests above](https://github.com/Baseflow/flutter-permission-handler/issues/1333#issuecomment-2539725309) are the same for me. 

Cheers

**Comment by Phenek** [comment]
Hello guys,

So you all have a workaround, even a fix!
That's pretty nice!

Would you please make a pull request?
It will help a lot, the community, not just me.

**Comment by Phenek** [comment]
### In fact this is not an issue, as the Readme says
You must list the permissions you want to use in your application:

## **Add the following to your Podfile file**:

Convert this:
````
post_install do |installer|
  installer.pods_project.targets.each do |target|
    flutter_additional_ios_build_settings(target)
  end
end

````

to this:
````
post_install do |installer|
  installer.pods_project.targets.each do |target|
    flutter_additional_ios_build_settings(target)

    target.build_configurations.each do |config|
      config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= [
        '$(inherited)',
    
        ## dart: PermissionGroup.bluetooth
        'PERMISSION_BLUETOOTH=1',
      ]
    end
  end
end
````
PS: The comment `##dart: PermissionGroup.bluetoot` is needed!!

You can close this issue, Thanks for your time

**Comment by Phenek** [comment]
### My workaround on Bluetooth permission using FlutterBluePlus:

```
  static Future<BluetoothAdapterState> getFirstBluetoothAdapterState() async {
    final completer = Completer<BluetoothAdapterState>();
    late StreamSubscription<BluetoothAdapterState> subscription;

    subscription = FlutterBluePlus.adapterState.listen((BluetoothAdapterState state) {
      if (state == BluetoothAdapterState.unknown) return;
      
      completer.complete(state);
      subscription.cancel();
    });

    return completer.future;
  }

  static Future<bool> isBluetoothEnabled() async {
    var state = await getFirstBluetoothAdapterState();
    if (state == BluetoothAdapterState.on) {
      return true;
    } else if (state == BluetoothAdapterState.off) {
      showBluetoothDisabledPopup();
      return false;
    } else if (state == BluetoothAdapterState.unauthorized) {
      showBluetoothUnauthorizedPopup();
      return false;
    }
    return false;
  }
```

**Comment by tilongzs** [comment]
Thank you @Phenek , this solution works.
On iOS 16.7.11, iPhone 8, Flutter (Channel stable, 3.35.2), the first call to `await Permission.bluetooth.request();` always immediately returns `PermissionStatus.permanentlyDenied`, and only then does the system display the Bluetooth permission dialog. It works normally on Android, and I don’t know why it doesn’t wait on iOS. The FlutterBluePlus solution doesn’t have this issue.
